### PR TITLE
z/OS Unix System Services enablement

### PR DESCRIPTION
**Pull request by OldNavi**
Hello all,

i've seen many attempts to enable durable task plugin on native z/OS unix system services, i had it running fine  almost for one year - merging base master branch with my changes locally at my site and now decided to offer it. Could you please have a look if this enablement changes can be merged in  the plugin codebase  or advise what can i amend/change  ?

### CODE REVIEW

**Comment by OldNavi** [comment]
Hi @jglick - is there any chance that this PR can be reviewed and accepted ?

**Comment by RamchandV** [comment]
@jglick  waiting for this change , please review, @OldNavi thankyou for helping the community.

**Comment by jglick** [comment]
Is this intended to supersede #28 and/or #76?

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 1]
```diff
@@ -195,9 +196,8 @@ public String getScript() {
         private transient long checkedTimestamp;
 
         private ShellController(FilePath ws) throws IOException, InterruptedException {
-            super(ws);
-        }
-
+                super(ws);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Please do not change whitespace on unrelated lines.

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 2]
```diff
@@ -180,6 +181,7 @@ protected FileMonitoringController(FilePath ws) throws IOException, InterruptedE
             CountingOutputStream cos = new CountingOutputStream(transcodedSink);
             try {
                 log.act(new WriteLog(lastLocation, new RemoteOutputStream(cos)));
+                sink.flush();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: What does this have to do with z/OS support?

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 3]
```diff
@@ -223,13 +225,15 @@ protected FileMonitoringController(FilePath ws) throws IOException, InterruptedE
         }
 
         /** Avoids excess round-tripping when reading status file. */
-        static class StatusCheck extends MasterToSlaveFileCallable<Integer> {
+        /* replaced static class STATUS_CHECK with anonymous to capture charset */
+        final MasterToSlaveFileCallable<Integer> STATUS_CHECK_INSTANCE = new MasterToSlaveFileCallable<Integer>() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: No, you may not remote an anonymous inner class.

**Comment by jglick** [comment]
To be clear, I did already try to test UTF-16 as an encoding on a particular `sh` step using an existing test container, which predictably resulted in complete junk. I did not yet try to create a _new_ test container natively set to UTF-16 and use it to just run regular stuff with ASCII output. Since no Pipeline devs have access to a z/OS environment to test anything interactively, much less in CI, that is the closest we could come.

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java (hunk 2 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: Well, this is due to weird IBM Java behavior on z/OS. Without flushing stream some of the job output isn't timely appears in a log.  No side effects on OpenJDK or Sun Java.

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java (hunk 1 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Sorry, will fix.

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java (hunk 3 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: Mmm... Okay, can i use the same approach as in getOutput 
```
       @Override public byte[] getOutput(FilePath workspace, Launcher launcher) throws IOException, InterruptedException {
            return getOutputFile(workspace).act(new MasterToSlaveFileCallable<byte[]>() {
                @Override public byte[] invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
```

**Comment by OldNavi** [comment]
A couple of notices why i had to encode script sh  to EBCDIC on the z/OS USS platform.  First thing - we couldn't rely on a system default encoding in z/OS Java because it doesn't reflect actual system encoding at all (by default z/OS IBM Java uses UTF8 BTW). Setting it with -Dfile.encoding=EBCDIC code page can avoid transcoding of script.sh but it is global setting and affects a lot of other things in a jenkins agent and plugins  behavior.  Technically there is a way to have a script.sh in ASCII charset on z/OS USS and execute it but this requires to call a z/OS chtag utility to tag script.sh as file required for transcoding ( https://www.ibm.com/support/knowledgecenter/SSLTBW_2.2.0/com.ibm.zos.v2r2.bpxa500/chtag.htm )
which again means that we need to create some script and execute beforehand. I'm using jenkins to build z/OS legacy code like PL/I and Assembler and another dark side of the moon is an output of programs running on MVS side of z/OS which can be called via script.sh but doesn't accept ASCII encoding in any way. 
So in general i've tried to isolate z/OS USS behavior from a other platforms. 
i've also adopted git-client-plugin to work on z/OS USS - same issues are the.

**Comment by OldNavi** [comment]
Some comments why setting encoding at sh DSL is not helping on z/OS and i have to force encoding inside plugin (1-st reason is because writing script is only UTF-8 )

So running simple pipeline on z/OS with sh step set to IBM1047 encoding 


We are getting the following output

It's clear that jenkins-result.txt can't be parsed because in STATUS_CHECK_INSTANCE  invocation we don't do transcoding. in pull request i've changed it to honor charset from FileMonitoringTask



i personally don't like a check i've wrote here - maybe it is better to change a behavior and honor sh encoding setting when writing a script.sh - but it is a bold move, as we aren't going to transcode an output but an input too
` 
  String encoding = os == OsType.ZOS ? "IBM1047" : null;
        if(encoding != null) charset(Charset.forName(encoding));
        ShellController c = new ShellController(ws);
.....
 shf.write(s, encoding==null ? "UTF-8" : encoding);
 shf.chmod(0755);`

**Comment by OldNavi** [comment]
For this PR to work it needs JEP-206 implemented so there is a prerequisite of Pipeline Nodes and Processes Plugin > 2.20

**Comment by jglick** [comment]
BTW it is pointless to work in this area until after #60 is merged.

**Comment by OldNavi** [comment]
Hi Jesse,

Do you know any timeline when #60 is going to be merged, so i'll be able to continue to adopt my changes on top of it? We have just finished adopting git cli client to work on z/OS USS and merged into master https://github.com/jenkinsci/git-client-plugin/pull/362 so half work of z/OS USS enablement for Jenkins is done. :-)

**Comment by RamchandV** [comment]
> Hi Jesse,
> Do you know any timeline when #60 is going to be merged, so i'll be able to continue to adopt my changes on top of it? We have just finished adopting git cli client to work on z/OS USS and merged into master jenkinsci/git-client-plugin#362 so half work of z/OS USS enablement for Jenkins is done. :-)

#60 is merged now

**Comment by OldNavi** [comment]
Work in progress. Not yet ready for merge.

**Comment by OldNavi** [comment]
I think - i've done. @jglick - could you please review once again ?

**Comment by RamchandV** [comment]
Will there be any merge, or any idea when this change will be merged/ released ?

**Comment by OldNavi** [comment]
@jglick - any chance to have a look into this ?

**Comment by RamchandV** [comment]
@dwnusbaum ,@jglick, @svanoort, @rsandell, @dwnusbaum and @abayer , can anyone look @ it , this is on hold for a long time

**Comment by OldNavi** [comment]
i've removed sink.flush() call as IFUC - we have switched from pull to push logs technique... tested on mainframe and don't see delays in a logging anymore.

@dwnusbaum - thanks for review and feedback. Can this PR be accepted now ?

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 4]
```diff
@@ -106,9 +108,18 @@ public String getScript() {
         if (script.isEmpty()) {
             listener.getLogger().println("Warning: was asked to run an empty script");
         }
+        OsType os = ws.act(new getOsType());
+        String scriptEncodingCharset = "UTF-8";
+        if(os == OsType.ZOS) {
+            String ibmEncoding = ws.act(new getIBMzOsEncoding());
```
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: done.

**Comment by OldNavi** [comment]
Something went wrong with build Jenkins instance - commit's can't be build - but inside log there are exceptions logged through checkout phase. 
`Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to ubuntu-jenkinsinfra9f1f20
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1741)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
		at hudson.remoting.Channel.call(Channel.java:955)
		at hudson.FilePath.act(FilePath.java:1071)
		at hudson.FilePath.act(FilePath.java:1060)
		at hudson.FilePath.mkdirs(FilePath.java:1245)
		at hudson.plugins.git.GitSCM.createClient(GitSCM.java:810)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1180)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
		at `

**Comment by OldNavi** [comment]
Is there any chance for this PR to be accepted and merged ?

**Comment by jglick** [comment]
> if there was a separate encoding parameter that could be set to UTF-16 or some other non-ascii-superset encoding, and then we could use https://github.com/jenkinsci/docker-fixtures to create a test using a container with that encoding

Discussed earlier in https://github.com/jenkinsci/durable-task-plugin/pull/80#issuecomment-415545608. Could be picked up by someone.

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 5]
```diff
@@ -106,9 +108,17 @@ public String getScript() {
         if (script.isEmpty()) {
             listener.getLogger().println("Warning: was asked to run an empty script");
         }
+        OsType os = ws.act(new getOsType());
+        String scriptEncodingCharset = "UTF-8";
+        if(os == OsType.ZOS) {
+            if(SYSTEM_DEFAULT_CHARSET.equals(getCharset())) {
+            // Setting default charset to IBM z/OS default EBCDIC charset on z/OS if no encoding specified on sh step
```
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Because we can't hardcode normally...  EBCDIC is a set of codepages where symbols mapping differs. Worst thing that in EBCDIC that affects such chars like '! [ ]'  Default code page is  IBM-1047. but z/OS system programmer may set any other, like IBM-1025.  And if for example i'll hardcode IBM-1047 code page, but system is configured in IBM-1025 i will get bad chars in a script itself and it will fail. 
As example 
#!/bin/bash  will turn to smth like #|/bin/bash
'[ -f myfile ];'   will turn to smth like 'p -f myfille y;'

So the logic implemented is following (on z/OS only)
1/ If encoding parameter is set on sh encoding:  step - use it for transcode script, script output and script exit status)
2/ If no encoding specified on sh encoding: step -  go to z/OS and ask what codepage is set and apply it for transcoding

[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Because we can't hardcode normally... EBCDIC is a set of codepages where symbols mapping differs. Worst thing that in EBCDIC that affects such chars like '! [ ]' Default code page is IBM-1047. but z/OS system programmer may set any other, like IBM-1025. And if for example i'll hardcode IBM-1047 code page, but system is configured in IBM-1025 i will get bad chars in a script itself and it will fail.
As example
#!/bin/bash will turn to smth like #|/bin/bash
'[ -f myfile ];' will turn to smth like 'p -f myfille y;'

So the logic implemented is following (on z/OS only)
1/ If encoding parameter is set on sh encoding: step - use it for transcode script, script output and script exit status)
2/ If no encoding specified on sh encoding: step - go to z/OS and ask what codepage is set and apply it for transcoding

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 6]
```diff
@@ -241,12 +245,17 @@ protected FileMonitoringController(FilePath ws) throws IOException, InterruptedE
...
+            StatusCheck(String charset) {
+                this.charset = charset;
+            }
             @Override
             @CheckForNull
             public Integer invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
                 if (f.exists() && f.length() > 0) {
                     try {
-                        String fileString = Files.readFirstLine(f, Charset.defaultCharset());
+                        Charset cs = transcodingCharset(charset);
```
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: Well, we do expect that IBM Java on z/OS will report a value of  ibm.system.encoding with getProperty() call.  If such property doesn't exist we will get NULL returned (that means we aren't on z/OS IBM as this property is unique for z/OS Java). And in this case all code flow will be absolutely the same as on ASCII platform, with specifying Charset.defaultCharset() in readFirstLine call.

[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: I am not following your logic. If we are not on z/OS then `getIBMzOsEncoding` is not even called. Therefore this part of the patch is changing a `status.txt` read call on Linux from using `Charset.defaultCharset()` to using `transcodingCharset(charset)`, which could be anything—probably an ASCII superset, but not necessarily.

[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: I misunderstood your question then. On platforms rather than z/OS - behavior doesn't change much at all...  it just uses a same logic as for getOutput().   it will read status file with encoding specified either by defaultCharset() for the platform encoding or from sh step encoding:  parameter.  This logic seems to be more consistent if i'm not mistaken. Indeed i can leave as it was, creating an additional check for z/OS platform and transcode status file only we are running on z/OS, np here. Just let me know if such change needed.

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: I just wanted to give an option to user to manually set, encoding of the script on z/OS platform regardless of what is set on a system. But i can revert it back to Jenkins logic where sh encoding: determines only an output encoding, no problems here.

**Comment by OldNavi** [comment]
@jglick - I refactored code just to be specific on z/OS platform only, trying to avoid changes which could potentially affect open platforms.  Can you have a look if it is okay now ?

**Comment by gorbadoc** [comment]
I, too, am waiting for this.  Any chance it can get merged in the next month or so?  

I am the front of a wave that will (hopefully) push a bunch of CI into Jenkins if I can show developers that all they need to do is commit something and their development branch will get built across a bunch of platforms.  This is already done on a limited basis for UNIX variants but is only set up for production builds.  

What can I do to help expedite this?  I used to be a contributor to Eclipse CDT and was a committer on coreboot.

**Comment by OldNavi** [comment]
Hi  Devin,

Is there any chance to have a progress on this PR ?

**Comment by dwnusbaum** [comment]
@OldNavi Sorry that I have not had a chance to look yet. I will try to review the PR in the next few weeks. Part of the problem is that I am simply not familiar with z/OS, so it is difficult for me to review the changes, and without any tests I am concerned that we will regress the new functionality. When I have time to review I want to try to create a test using a Docker fixture for a container with a non-ASCII superset encoding (similar to tests like [this one](https://github.com/jenkinsci/durable-task-plugin/blob/d6258dd63384de5ec261e7374028fd6f4a106dd0/src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java#L309)) so that we can have automated tests for at least some of the behavior.

**Comment by gorbadoc** [comment]
I have a z/OS system available to me.  I have Ivan's patch installed and it
works as expected.  Perhaps I can run a test for you if you outline the
test that needs to be run?

I now have a monster makefile with recipes for several languages and am
successfully compiling, assembling, and linking several hundred modules
successfully in a multibranch pipeline.  I have two branches running and
will shortly, maybe today, expand to low tens of branches.

Does that help?

**Comment by OldNavi** [comment]
>  and without any tests I am concerned that we will regress the new functionality. When I have time to review I want to try to create a test using a Docker fixture for a container with a non-ASCII superset encoding...
I've tried to wrap z/OS logic just not to interfere with open systems at all. As for testing, I was thinking about "emulating" z/OS behavior with Docker - but it seems to be not trivial as for open systems. Playing with System.setProperty("file.encoding", "IBM1047") in runtime has no effect, as it should be only set up at startup time with -Dfile.encoding...  but setting that at start up time, completely breaks Jenkins agent behavior...
In general z/OS Java setup in z/OS is kind of mixed behavior in a difference for open systems... JVM itself runs with ASCII charset specified either UTF-8 or ISO8859-1 (which bidi maps to IBM1047).  Shell scripts need to be in EBCDIC code pages (IBM1047 by default) so we do a transcoding... Shell return code and output also comes in EBCDIC charset.

**Comment by hogstrom** [comment]
@dwnusbaum Just a data point.  I have a z/OS system on which I installed this fix and it correctly addressed the problem and I was able to build.  I'm aware of a large number of z/OS customers that are incorporating Jenkins to improve their CI/CD pipeline and this is a needed fix for them.  

I understand that you do not have access to a z/OS system.  I offer to get you access to one and to take you through the system to improve your familiarity of the problem and how the platform works.  I'm working on Project [Zowe](https://github.com/zowe/) and we need this functionality to build open source software on Z as well.  The system is on the Internet and we can spend whatever time is needed via WebEx if that would be a benefit to you.  Contact information is in my Git Profile.  

Thanks for your consideration on bringing this issues to a close.

@OldNavi Thanks for the work here ... very much appreciated.

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 7]
```diff
@@ -114,6 +114,11 @@ protected FileMonitoringController launchWithCookie(FilePath workspace, Launcher
         charset = SYSTEM_DEFAULT_CHARSET;
     }
 
+    @CheckForNull public final String getCharset()
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: No need to expose this in the API.
```suggestion
    @CheckForNull final String getCharset()
```
(or just remove `private` from the field)

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 8]
```diff
@@ -161,6 +166,10 @@ protected FileMonitoringController doLaunch(FilePath workspace, Launcher launche
         /** @see FileMonitoringTask#charset */
         private @CheckForNull String charset;
 
+        protected String getCharset() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: Ditto.
```suggestion
        String getCharset() {
```
(or just remove `private` from the field)

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 9]
```diff
@@ -210,7 +228,16 @@ private FilePath pidFile(FilePath ws) throws IOException, InterruptedException {
         }
 
         @Override protected Integer exitStatus(FilePath workspace, TaskListener listener) throws IOException, InterruptedException {
-            Integer status = super.exitStatus(workspace, listener);
+            Integer status = null;
+            OsType os = workspace.act(new getOsType());
```
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: (Note that deserialized copies of the controller will receive the field uninitialized, i.e. `false`, which is what we want.)

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java (hunk 8 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: Done.

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Done

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 10]
```diff
@@ -210,7 +228,16 @@ private FilePath pidFile(FilePath ws) throws IOException, InterruptedException {
         }
 
         @Override protected Integer exitStatus(FilePath workspace, TaskListener listener) throws IOException, InterruptedException {
-            Integer status = super.exitStatus(workspace, listener);
+            Integer status = null;
```
[the remark this answers is not in the document]
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: Done

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 11]
```diff
@@ -122,13 +122,14 @@ public String getScript() {
...
+                // Setting default charset to IBM z/OS default EBCDIC charset on z/OS if no encoding specified on sh step
+                charset(zOSSystemEncodingCharset);
             }
-            scriptEncodingCharset = zOSSystemEncodingCharset != null ? zOSSystemEncodingCharset.name() : scriptEncodingCharset;
+            scriptEncodingCharset = zOSSystemEncodingCharset.name();
         }
-
+        
         ShellController c = new ShellController(ws);
+        c.isZos = (os == OsType.ZOS);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: For stylistic purposes, nicer to make the field `final` and move its initialization to the constructor.

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java [hunk 12]
```diff
@@ -526,4 +535,4 @@ private synchronized static ScheduledExecutorService watchService() {
 
     }
 
-}
+}
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: revert (want final newlines)

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java (hunk 11 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: done

File: src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java (hunk 12 above)
Comment (OldNavi) [reply src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java]: done

File: src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java [hunk 13]
```diff
@@ -216,10 +215,15 @@ public String getScript() {
         private transient long checkedTimestamp;
 
         /** Caching zOS flag to avoid round trip calls in exitStatus()         */
-        private boolean isZos;
-
+        private final boolean isZos;
+        
         private ShellController(FilePath ws) throws IOException, InterruptedException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java]: You can delete the unused overload.

**Comment by hogstrom** [comment]
@dwnusbaum Devin, as part of the Zowe project that is building z/OS Open Source we are creating instances to be used by open source projects for build testing.  Although a small portion of the community, if we could help to provide and maintain an instance for the Jenkins community we'd be interested in helping out.  Please let me know if there is interest in including this platform as part of the Jenkins testing.

**Comment by dwnusbaum** [comment]
@hogstrom I appreciate your offer, and perhaps someone else on this PR would be interested in looking into it, but I am not able to commit to driving that effort myself.

**Comment by OldNavi** [comment]
> I just ran into this issue and found that the fix does NOT work on intel Jenkins servers. Just Unix based systems....

@nlopez59 - can you share more details ?

**Comment by gin-nader** [comment]
I'm hitting an issue where the `jenkins-result.txt` file is getting tagged as ASCII on a z/OS machine. This is causing the `StatusCheckWithEncoding(getCharset())`  check at https://github.com/jenkinsci/durable-task-plugin/blob/0d250a6142908f591bda91d8658ab1b102611ce3/src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java#L298 to throw an Exception.

Should I open a PR for the fix?
